ui/src/turing/components/form/standard_ensembler/LinkedRoutesTable.js

```diff
@@ -0,0 +1,116 @@
+
+  // this stringified value of routes below allows the React effect below to mimic a deep comparison when changes to the
+  // array routes are made
+  const stringifiedRoutes = JSON.stringify(routes)
+
+  // reset loaded routeToExperimentMappings if treatmentConfigRouteNamePath or routes changes
+  useEffect(() => {
+    if (isAllExperimentsLoaded) {
+      let newRouteToExperimentMappings = routes.reduce((m, r) => {m[r.id] = {running: {}, scheduled: {}}; return m}, {});
+      for (let experiment of allExperiments) {
+        for (let treatment of experiment.treatments) {
+          let configRouteName = getRouteName(treatment.configuration, treatmentConfigRouteNamePath);
+          if (typeof configRouteName === 'string' && configRouteName in newRouteToExperimentMappings) {
+            newRouteToExperimentMappings[configRouteName][getExperimentStatus(experiment).label.toLowerCase()][experiment.id] = experiment;
+          }
+        }
+      }
+      setRouteToExperimentMappings(newRouteToExperimentMappings);
+    }
+  }, [treatmentConfigRouteNamePath, stringifiedRoutes, routes, isAllExperimentsLoaded, allExperiments]);
+
+  const columns = [
+    {
+      field: "id",
+      width: "5px",
+      render: (id) => {
+        const isAssigned = routeToExperimentMappings[id] ?
+          Object.keys(routeToExperimentMappings[id].running).length +
+          Object.keys(routeToExperimentMappings[id].scheduled).length > 0 : false;
+        return (
+          <EuiIcon
+            type={isAssigned ? "check" : "cross"}
+            color={isAssigned ? "success" : "danger"}
+            size={"m"}
+            style={{ verticalAlign: "sub" }}
+          />
+        );
+      },
+    },
+    {
+      field: "id",
+      width: "20%",
+      name: "Route Name",
+      render: (id) => {
+        const isAssigned = routeToExperimentMappings[id] ?
+          Object.keys(routeToExperimentMappings[id].running).length +
+          Object.keys(routeToExperimentMappings[id].scheduled).length > 0 : false;
+        return (<EuiTextColor color={isAssigned ? "success" : "danger"}>{id}</EuiTextColor>);
+      },
+    },
+    {
+      field: "id",
+      width: "35%",
+      name: "Running Experiments",
+      render: (id) => (
+        <LinkedExperimentsContextMenu
+          projectId={projectId}
+          linkedExperiments={routeToExperimentMappings[id]}
+          experimentStatus={"running"}
+        />
+      ),
+    },
+    {
+      field: "id",
+      width: "35%",
+      name: "Scheduled Experiments",
+      render: (id) => (
+        <LinkedExperimentsContextMenu
+          projectId={projectId}
+          linkedExperiments={routeToExperimentMappings[id]}
+          experimentStatus={"scheduled"}
+        />
+      )
+    },
+  ];
+
+  return isAllExperimentsLoaded ? (
+    <EuiFlexItem>
+      <EuiInMemoryTable
+        items={routes.filter((r) => r.id !== "")}
```

Comment:
In what cases, would `r.id` be an empty string?

Reply:
When a user clicks on `+ Add Route` and then doesn't type anything in the route name 👀 Or when he/she erases everything he/she has typed for a route name.